extensions/optaplanner/deployment/src/main/java/io/quarkus/optaplanner/deployment/OptaPlannerProcessor.java

```diff
@@ -0,0 +1,250 @@
+
+import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
+import io.quarkus.arc.deployment.BeanContainerListenerBuildItem;
+import io.quarkus.deployment.Capabilities;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.annotations.Record;
+import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.RuntimeInitializedClassBuildItem;
+import io.quarkus.deployment.recording.RecorderContext;
+import io.quarkus.optaplanner.OptaPlannerBeanProvider;
+import io.quarkus.optaplanner.OptaPlannerObjectMapperCustomizer;
+import io.quarkus.optaplanner.OptaPlannerRecorder;
+
+class OptaPlannerProcessor {
+
+    OptaPlannerQuarkusConfig optaPlannerQuarkusConfig;
+
+    @BuildStep
+    FeatureBuildItem feature() {
+        return new FeatureBuildItem(FeatureBuildItem.OPTAPLANNER);
+    }
+
+    @BuildStep
+    void registerAdditionalBeans(BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
+        // The bean encapsulating the SolverFactory
+        additionalBeans.produce(new AdditionalBeanBuildItem(OptaPlannerBeanProvider.class));
+    }
+
+    @BuildStep(loadsApplicationClasses = true)
+    @Record(STATIC_INIT)
+    void recordSolverFactory(OptaPlannerRecorder recorder, RecorderContext recorderContext,
+            CombinedIndexBuildItem combinedIndex,
+            BuildProducer<ReflectiveHierarchyBuildItem> reflectiveHierarchyClass,
+            BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
+            BuildProducer<BeanContainerListenerBuildItem> beanContainerListener) {
+        ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+        SolverConfig solverConfig;
+        if (optaPlannerQuarkusConfig.solverConfigXml.isPresent()) {
+            String solverConfigXML = optaPlannerQuarkusConfig.solverConfigXml.get();
+            if (classLoader.getResource(solverConfigXML) == null) {
+                throw new IllegalStateException("Invalid optaplanner.solverConfigXML property (" + solverConfigXML
+                        + "): that classpath resource does not exist.");
+            }
+            solverConfig = SolverConfig.createFromXmlResource(solverConfigXML, classLoader);
+        } else if (classLoader.getResource(OptaPlannerQuarkusConfig.DEFAULT_SOLVER_CONFIG_URL) != null) {
+            solverConfig = SolverConfig.createFromXmlResource(
+                    OptaPlannerQuarkusConfig.DEFAULT_SOLVER_CONFIG_URL, classLoader);
+        } else {
+            solverConfig = new SolverConfig(classLoader);
+        }
+        // The deployment classLoader must not escape to runtime
+        solverConfig.setClassLoader(null);
```

Comment:
Just out of curiosity, what CL will be used at runtime as a result of this?

Reply:
Most of the time, it's not used, except if a `scoreDRL` is active.
If it's `null` in that case, `ScoreDirectorFactoryConfig` does this at runtime, during `buildSolver()`:

```
ClassLoader actualClassLoader = (classLoader != null) ? classLoader : getClass().getClassLoader();
```

Once we start doing more work during the static init, after we internalize the SolverConfig in the SolverFactory (currently impossible due to deprecated methods that will dissappear in 8.0), I 'll have to come back to this, as that code might become a problem then.
Good question indeed. :)

Reply:
What am I sort of worried about is the dev-mode. There are certain situations where `getClass().getClassLoader()` might not give you what you expect.
Perhaps also adding `QuarkusDevModeTest` is a good idea to make sure things work properly?

Reply:
I 've been trying to add in QuarkusDevModeTest, but I reluctant to add a test that doesn't work in IntelliJ (it hurts productivity). In IntelliJ I get the "Undeclared build item class io.quarkus.deployment.builditem.ApplicationClassNameBuildItem" exception, but on the command line, I don't.

Reply:
IntelliJ can mess up the classpath when running tests inside the Quarkus repo. I don't think it's something we need to worry about

Reply:
The dev mode test should prove that things are OK with this change. 
To be 100% Quarkus compliant, upstream here should provide a flag to use the TCCL when the classloader is null, but I don't think we need it now

Reply:
Using ThreadContextClassLoader is what it does by default if we don't set it to null.
When I remove the solverConfig.setClassLoader(null); line, I get

```
Caused by: java.lang.RuntimeException: Unable to serialize objects of type class io.quarkus.bootstrap.classloading.QuarkusClassLoader to bytecode as it has no default constructor
        at io.quarkus.deployment.recording.BytecodeRecorderImpl$53.createValue(BytecodeRecorderImpl.java:1273)
```